pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/GrpcBrokerRequestHandler.java

```diff
@@ -0,0 +1,198 @@
+import org.apache.pinot.spi.env.PinotConfiguration;
+import org.apache.pinot.spi.utils.builder.TableNameBuilder;
+
+
+/**
+ * The <code>SingleConnectionBrokerRequestHandler</code> class is a thread-safe broker request handler using a single
+ * connection per server to route the queries.
+ */
+@ThreadSafe
+public class GrpcBrokerRequestHandler extends BaseBrokerRequestHandler {
+  private static final int DEFAULT_MAXIMUM_REQUEST_ATTEMPT = 10;
+
+  private final GrpcQueryClient.Config _grpcConfig;
+  private final Map<String, AtomicInteger> _concurrentQueriesCountMap = new ConcurrentHashMap<>();
+  private final int _maxBacklogPerServer;
+  private final int _grpcBrokerThreadPoolSize;
+  private final StreamingReduceService _streamingReduceService;
+
+  private transient PinotStreamingQueryClient _streamingQueryClient;
+
+  public GrpcBrokerRequestHandler(PinotConfiguration config, RoutingManager routingManager,
+      AccessControlFactory accessControlFactory, QueryQuotaManager queryQuotaManager, TableCache tableCache,
+      BrokerMetrics brokerMetrics, TlsConfig tlsConfig) {
+    super(config, routingManager, accessControlFactory, queryQuotaManager, tableCache, brokerMetrics);
+    _grpcConfig = buildGrpcQueryClientConfig(config);
+
+    // parse and fill constants
+    _maxBacklogPerServer = Integer.parseInt(_config.getProperty("MAXIMUM_BACKLOG_PER_SERVER", "1000"));
+    _grpcBrokerThreadPoolSize = Integer.parseInt(_config.getProperty("GRPC_BROKER_THREAD_POOL_SIZE", "10"));
+
+    // create streaming query client
+    _streamingQueryClient = new PinotStreamingQueryClient(_grpcConfig);
+
+    // create streaming reduce service
+    _streamingReduceService = new StreamingReduceService(config);
+  }
+
+  @Override
+  public void start() {
+  }
+
+  @Override
+  public synchronized void shutDown() {
+    _streamingReduceService.shutDown();
+  }
+
+  @Override
+  protected BrokerResponseNative processBrokerRequest(long requestId, BrokerRequest originalBrokerRequest,
+      @Nullable BrokerRequest offlineBrokerRequest, @Nullable Map<ServerInstance, List<String>> offlineRoutingTable,
+      @Nullable BrokerRequest realtimeBrokerRequest, @Nullable Map<ServerInstance, List<String>> realtimeRoutingTable,
+      long timeoutMs, ServerStats serverStats, RequestStatistics requestStatistics)
+      throws Exception {
+    assert offlineBrokerRequest != null || realtimeBrokerRequest != null;
+
+    String rawTableName = TableNameBuilder.extractRawTableName(originalBrokerRequest.getQuerySource().getTableName());
+    Map<ServerRoutingInstance, Iterator<Server.ServerResponse>> responseMap = new HashMap<>();
+    // Making request to a streaming server response.
+
+    if (offlineBrokerRequest != null) {
+      // to offline servers.
+      assert offlineRoutingTable != null;
+      Map<ServerRoutingInstance, Iterator<Server.ServerResponse>> serverResponses =
+          streamingQueryToPinotServer(requestId, _brokerId, rawTableName, TableType.OFFLINE,
```

Comment:
We can reduce the garbage created by passing in the responseMap and directly adding entries to it